### ISSUE TITLE
investigate whether vmovdqa is worth 32-byte alignment

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#438]
_From [bruen...@google.com](https://code.google.com/u/109494838902877177630/) on April 20, 2011 11:17:43_

split from issue #433 + performance of non-32-byte alignment: once have SandyBridge processor
  need to measure cost of vmovdqu and whether worth arranging 32-byte
  alignment for fcache cxt switch and/or clean call

_Original issue: http://code.google.com/p/dynamorio/issues/detail?id=438_

**Comment by hgreving2304** [comment]
I am measuring the following CPIs

AVX loads (ymm)

vmovdqa
4.018271179
vmovdqu <aligned>
4.014084095
vmovdqu <unaligned>
8.221283677
vmovdqu <unaligned, cross cache line>
8.220376073

AVX stores (ymm)

vmovdqa
8.120018776
vmovdqu <aligned>
8.139037985
vmovdqu <unaligned>
16.01327721
vmovdqu <unaligned, cross cache line>
16.00491806

**Comment by derekbruening** [comment]
I think you mis-labeled some of those identical `vmovdqu`?

**Comment by hgreving2304** [comment]
No, the text in "<>" wasn't visible. Changed into "[]". It's vmovdqu with aligned address, unaligned address and unaligned address + 2 cache lines.

**Comment by derekbruening** [comment]
It would be best to include the microbenchmark code, as was done for most experiments above, for completeness.

**Comment by hgreving2304** [comment]
This is the .s output of the main loops of the benchmark. The pointer of the main buffer is 64-byte aligned and is offset by 1 (unaligned) or 33 (unaligned + cross cache line)

Main loop load + vmovda

`       
salq    $32, %rdx
        movl    %eax, %eax
        movl    $10000000, %ecx
        movq    %rdx, %rdi
        leaq    8192(%rsi), %rdx
        orq     %rax, %rdi
        .p2align 4,,10
        .p2align 3
.L7:
        movq    %rsi, %rax
        .p2align 4,,10
        .p2align 3
.L10:
#APP
# 62 "measure_alignment.cpp" 1
        vmovdqa (%rax), %ymm0
vmovdqa 32(%rax), %ymm1
vmovdqa 64(%rax), %ymm2
vmovdqa 96(%rax), %ymm3
vmovdqa 128(%rax), %ymm4
vmovdqa 160(%rax), %ymm5
vmovdqa 192(%rax), %ymm6
vmovdqa 224(%rax), %ymm7

# 0 "" 2
#NO_APP
        addq    $256, %rax
        cmpq    %rdx, %rax
        jne     .L10
        subl    $1, %ecx
        jne     .L7
`

Main loop load + vmovdqu [unaligned]

`
        salq    $32, %rdx
        movl    %eax, %eax
        leaq    1(%r8), %rsi
        movq    %rdx, %rdi
        leaq    8193(%r8), %rdx
        movl    $10000000, %ecx
        orq     %rax, %rdi
        .p2align 4,,10
        .p2align 3
.L7:
        movq    %rsi, %rax
        .p2align 4,,10
        .p2align 3
.L10:
#APP
# 62 "measure_alignment.cpp" 1
        vmovdqu (%rax), %ymm0
vmovdqu 32(%rax), %ymm1
vmovdqu 64(%rax), %ymm2
vmovdqu 96(%rax), %ymm3
vmovdqu 128(%rax), %ymm4
vmovdqu 160(%rax), %ymm5
vmovdqu 192(%rax), %ymm6
vmovdqu 224(%rax), %ymm7

# 0 "" 2
#NO_APP
        addq    $256, %rax
        cmpq    %rdx, %rax
        jne     .L10
        subl    $1, %ecx
        jne     .L7
`

Main loop load + vmovdqu [unaligned, cross cache line]

`
        salq    $32, %rdx
        movl    %eax, %eax
        leaq    33(%r8), %rsi
        movq    %rdx, %rdi
        leaq    8225(%r8), %rdx
        movl    $10000000, %ecx
        orq     %rax, %rdi
        .p2align 4,,10
        .p2align 3
.L7:
        movq    %rsi, %rax
        .p2align 4,,10
        .p2align 3
.L10:
#APP
# 62 "measure_alignment.cpp" 1
        vmovdqu (%rax), %ymm0
vmovdqu 32(%rax), %ymm1
vmovdqu 64(%rax), %ymm2
vmovdqu 96(%rax), %ymm3
vmovdqu 128(%rax), %ymm4
vmovdqu 160(%rax), %ymm5
vmovdqu 192(%rax), %ymm6
vmovdqu 224(%rax), %ymm7

# 0 "" 2
#NO_APP
        addq    $256, %rax
        cmpq    %rdx, %rax
        jne     .L10
        subl    $1, %ecx
        jne     .L7
`

**Comment by hgreving2304** [comment]
According to the title of this issue and my measurements, the answer would be yes. Any other ideas/comments? In steady state, context switches shouldn't happen very often, same for reading signal frames?

**Comment by derekbruening** [comment]
> According to the title of this issue and my measurements, the answer would be yes.

It looks like all the numbers, both the old and new, show quite small differences, right?  So while the aligned version is definitely not slower, it is only marginally faster, so if it were a real pain to ensure alignment, the answer might not necessarily be an emphatic yes that we should use the aligned.  If it's easy to align, then we should do it.

**Comment by hgreving2304** [comment]
> So while the aligned version is definitely not slower, it is only marginally faster

I measured twice as fast! What did not make a difference was whether the unaligned version access 1 or 2 cache lines.

Maybe we cross-check with exegesis results?

**Comment by derekbruening** [comment]
Sorry, yes, aligned vs unaligned is a big difference, and aligning for the majority of cases is worthwhile.  What is in my head is whether it's worth using the aligned opcode in shared code that is sometimes not aligned, when it's a pain to ensure it's aligned -- IIRC there was something like that in DR but it is definitely worth re-investigating.  That is what I meant to say but was not clear about: that if a small % of uses are unaligned and it's painful to get them aligned, using the unaligned opcode seems not unreasonable.

**Comment by derekbruening** [comment]
Yes, agreeing we have the data this issue is about.

**Comment by hgreving2304** [comment]
xref #1312

**Comment by hgreving2304** [comment]
Closing this issue, I think we got what we wanted.

**Comment by hgreving2304** [comment]
Actually leaving this open, since it now addresses whether it is worth _in DR_.
